winc.man



WINC(1)							  WINC(1)


NAME
       winc  -	command	 line interface to the WinCam.One digital
       camera

SYNOPSIS
       winc [-options]

DESCRIPTION
       winc gives UNIX shell access to the digital  camera  known
       as the WinCam.  It is a 640x480 capable color digital cam-
       era, interfaced to a computer via  a  serial  cable.   See
       http://www.wincam.com for more information about this cam-
       era.

       Images grabbed by winc are dumped in PGM or PNM format  to
       the  standard  output, with the intention that they be run
       through other tools to be converted to the desired format,
       e.g.:

       winc -g -2 | cjpeg >image.jpg

       (The cjpeg command referred to is available as part of the
       Independent JPEG Group's distribution  of  jpeg	software.
       The PGM, PNM, and PBM formats are specified as part of the
       widely available netpbm utilities.  Both of these packages
       are  probably  available	 at  the same place you found the
       winc code.)
      
       (OS/2 Note:  The Independent JPEG Group's JPEG Software v 6a 
       is currently at: 
           ftp://hobbes.nmsu.edu/old/os2/graphics/jpeg6a_os2a.zip

       The GBM Utils also deal with the PNM format - they can 
       currently be found at:
	   ftp://hobbes.nmsu.edu/old/os2/graphics/gbm.zip

       As the Hobbes archive is re-organizing as this is being written,
       expect this to change. )
	
       Although this software processes the full 492 lines avail-
       able  to	 the  camera,  there may be some residual effects
       from the CCD (Charge Coupled Device) in the camera visible
       in  the	first  and last few lines.  Hence the claimed 480
       line resolution.

OPTIONS
       There are far too many  options	to  this  program,  as	a
       result  of  its	being  a  development  tool  for the winc
       library in addition to the basic	 image	capture	 command.
       The  important  options	to  know about are -c (color), -g
       (greyscale),  -2	 (double  scan),  -z   (zoom/scale),   -b
       (brightness),  and  -l  (lighting).   The image will go to
       stdout by default.  winc will  do  nothing  useful  except
       find the camera (i.e. no image will be snapped) unless one
       of -c, -g, or -v is present.

       -h     prints the short usage message.

       -H     prints the full usage message, since it's so  long.

       -c     Grabs  a	color  image.	Color  images are 24 bits
	      deep, with 255 values for each of red,  green,  and
	      blue.   The  default size is 640x492.  Output is in
	      PNM format.

       -g     Grabs a greyscale image.	Greyscale  images  are	8
	      bits  deep,  with a maximum grey level of 255.  The
	      default size is 640 by 492.  Output is in PGM  for-
	      mat.

       -v     captures	a  "viewfinder" image.	Viewfinder images
	      are 64 by 48 greyscale imagemaps, with grey  values
	      in  the  range 0 to 15.  Because they are small and
	      "shallow", they take much	 less  time  to	 download
	      from the camera.	Output is in PGM format.

       -2     Normally	images	are captured in a single scan: in
	      this mode, of the 492 lines  in  a  default  image,
	      only  the	 even  rows are "real";	 the odd rows are
	      created by averaging between  the	 rows  above  and
	      below.   If the subject matter is still enough, the
	      -2 option can be used to force a second scan  about
	      1/8th of a second later.	This will fill in the odd
	      rows with real data.

       -z N   Sets the viewfinder zoom factor to 1, 2, 4,  or  8,
	      or  for  full images, sets the full image reduction
	      factor.  Viewfinder images normally show	the  full
	      field  of	 view at low resolution, but specifying a
	      zoom of 8 will yield the camera's	 maximum  resolu-
	      tion  in	a relatively small area.  This can assist
	      in focusing the camera lens  (which  must	 be  done
	      manually	--  loosen  the	 screw!).  Regular images
	      cannot be zoomed, but can instead have their  reso-
	      lution  and  overall dimension reduced.  This might
	      be done when the full imagesize  or  resolution  is
	      not  wanted  or  needed.	The full field of view is
	      preserved.  The values 1, 2,  4,	and  8	represent
	      linear  scaling of both the horizontal and vertical
	      axes.  Thus "-z 2" gives an image that  is  160x123
	      rather than the default 640x492.	(So actually, the
	      arguments should really be 1, 4, 16, and 64.)

       -l lighting
	      Affect the color balance of the image.   See  COLOR
	      BALANCE below.

       -b NN  Sets  the	 "brightness" goal for the camera.  Expo-
	      sure will be adjusted automatically in  an  attempt
	      to  reach	 this  percentage  of  full  white.   The
	      default value is 50 (percent).

       -S NN  Saves a the threshold above which a pixel	 is  con-
	      sidered  "bad".  See the section below on STARFIELD
	      CORRECTION for more information.

       -e NN or -E NN
	      Sets exposure in milliseconds(-e)	 or  microseconds
	      (-E).   The  shortest  possible  exposure	 time  is
	      1/100,000th of a second, the longest  is	a  little
	      over  1.3 seconds.  If neither option is used, winc
	      will try several exposures until the average  pixel
	      brightness goal is reached (see -b).

       -a     Sound an audible alarm when the picture is actually
	      snapped.	The process of "finding" the camera, tak-
	      ing  the	picture,  and  downloading  the	 image is
	      fairly lengthy.  Hearing a small beep when the pic-
	      ture has been taken can be helpful.

       -o FILE
	      Normally	all  output  from winc will go to stdout.
	      This will cause output to go to the specified file-
	      name instead.  If -o is used, then the modification
	      time of "filename" will be set to the time at which
	      the  picture  was snapped, rather than being set by
	      the system to the time at which the  last	 part  of
	      the  file was written.  (The time difference can be
	      tens of seconds.)

       -i IDSTRING
	      Specify the "name" of the camera to be used.   This
	      name  will  affect  the  way in which configuration
	      information will be looked up.  See the section  on
	      MULIPLE CAMERAS, below.

       -n     Leave  the image "narrow".  The natural image taken
	      by the camera looks tall and thin if viewed unmodi-
	      fied.  Normally a horizontal scaling is applied, to
	      bring it up to 640 wide, which makes circles  round
	      again.  Using -n suppresses this widening.  If this
	      is  done,	 the  default  size  of	 images	 will  be
	      512x492.

       -r     Leave  a raw image, rather than one with brightened
	      corners.	The camera leaves the corners quite  dark
	      (half  as bright as the center), and winc will nor-
	      mally brighten them.  Using -r suppresses this.

       -G VAL Runs a gamma function on the  image,  which  is  an
	      exponential curve that affects some areas more than
	      others.  A value of 100 (or 0) will leave the image
	      unchanged.   Values  less than 100 will tend to dim
	      the image, values greater than  100  will	 tend  to
	      brighten	it.   The  default  value  is  110, which
	      lightens shadowy areas.  The default value  is  set
	      with the WinCamGamma variable.

       -d     Runs  camera diagnostics.	 If combined with -g, the
	      camera will download as an image the  pattern  left
	      in  memory as a result of the diagnostics, and this
	      can be viewed as a test pattern.

       -t tracestring
	      Turns on program tracing.	 The argument is a string
	      of  consisting of traceflags.  Primarily for debug-
	      ging, this can be used to track down  configuration
	      problems of various sorts.  The possible flags are:
		   v just shows program and camera version information
		   V will turn on all possible messages
		   d for diagnostic tracing
		   e to get exposure information
		   a for simple API tracing
		   A for full API tracing
		   p for camera protocol
		   P for _all_ bytes of protocol
		   c for serial port control
		   l for camera lock/unlock tracing
		   g for graphics output routine tracing
		   r for winc.rc configuration file tracing
		   k for color processing tracing
		   x for program-level tracing (i.e outside the library)

       -f     The winc library supports a couple of ways of  tak-
	      ing  pictures.   Normally,  picture  processing  is
	      overlapped with the download of data from the  cam-
	      era.   The -f option causes a different set of rou-
	      tines to be used, which first  download,	and  then
	      process the image.

       -R     Data  retrieved  from the camera is dumped in Star-
	      Dot's .wcd format to stdout.

       -R -R  Data retrieved from the camera is	 dumped	 with  no
	      processing or conversion to stdout.

       -D     Raw  image  data	in  StarDot's .wcd format is read
	      from stdin.  The data is always  considered  to  be
	      double  scan,  so	 effectively  a	 '-2'  is forced.
	      Fractional scans are not supported, so the -z  flag
	      cannot be used.  The -g and -c flags are honored.

       -D -D  This   will   cause  raw	camera	data  (presumably
	      obtained with the "-R -R" option) to be  read  from
	      stdin,  and processed as usual.  The -z, -c, and -g
	      flags are still honored.

COLOR BALANCE
       Either the -l option or the WinCamColor	variable  can  be
       used  to	 specify the lighting of the scene being snapped,
       to affect the "color balance" of the  camera.   The  value
       can  be	one  of	 several  predefined strings ("sunlight",
       "flourescent", "tungsten" (or  the  synonymous  "incandes-
       cent"),	"halogen", "studio", "filter", "defaults"), or it
       can be a parenthesized triple of percentages by	which  to
       adjust  the  red, green, and blue components of the image,
       respectively.  Default values are obtained by omitting  -l
       entirely,  or  using  "-l defaults" or "-l (100,100,100)".
       (The parentheses will need to be quoted from the shell.)

       Aliases for particular lighting needs can  be  set  up  by
       defining environment or winc.rc (see CONFIGURATION, below)
       variables  of  the  form	  "WinCamColor_ALIASNAME".    For
       instance,       one	 might	     define	 "WinCam-
       Color_backyard=(120,120,100)", and then refer to	 it  with
       "-l  backyard".	Aliases can also refer to the compiled-in
       values, e.g.  "WinCamColor_office=flourescent".

STARFIELD CORRECTION
       Like most CCD cameras, the WinCam will  have  some  pixels
       that appear very bright when taking a picture in low light
       conditions.  The effect when looking at a  dark	image  is
       that it is speckled, or has a starry appearance.	 The good
       news is that the "bad" CCD sensors are fixed for	 a  given
       camera.	 By  saving  a	reference  image,  the "starfield
       effect" can be eliminated from later pictures.

       Use the -S option to create a reference image.	The  cam-
       era's  lens  cap	 should	 be left on, so that it takes the
       snap "in the dark".  The output is in the  PBM  monochrome
       bitmap format.  The parameter to -S is the threshold above
       which a pixel is considered "bad".  Values  of  20  to  50
       seem  right,  apparently	 depending on the camera.  It may
       also  depend  on	 the  temperature  of  the  camera.   The
       starfield  image	 can  be viewed with a program like xv to
       see if it seems to have too many or too	few  "bad"  spots
       recorded.  Test the starfield image by using it to correct
       another normal image, also taken with the lens cap left in
       place.	The  resulting	image should be (almost) entirely
       dark, or at least much better than a similar  image  taken
       without starfield correction.

       The  name  of  the file used for the reference is found in
       the WinCamStarfield configuration variable.   The  easiest
       way  to	test a new reference is to use a command like the
       following:

       WinCamStarfield=starfield.pbm  winc -c -ta | xv -

       (assuming the reference	was  saved  in	"starfield.pbm").

ChangeLog

Changes for winc-36:
    961016
	+ README: mention the mailing list.  subscribe at:
	    pgf-winc-request@foxharp.boston.ma.us
    960929
	+ winc.c wincd.c xwinc.c: added missing break in getopt switch.

Changes for winc-35:
    960926
	+ find.c: zero the cam struct after mallocing.  current symptom
	  was an uninitialized "starfield" element that was causing
	  core dumps.

	+ wincd.c wincd.1: add -e and -E for controlling exposure.

Changes for winc-34:
    960910
	+ wincd.c winc.1: apply gamma function, update man page.

	+ vignette.c winc.c wincam.h:  implemented winc_gamma(), which
	  applies a gamma function to the image.  the default gamma value
	  is 100 (representing 1.0 in the actual equation).

    960828
	+ xwinc.1 winc.1: clarifications, syntax errors.

	+ view.c: if avg becomes zero while auto-exposing, set to high
	  default, since it might be zero due to not enough light.  this
	  is probably only the case when the lens cap is on, but possibly
	  at night?

	+ winc.c: choose a goal when doing starfield shots.  let's call it
	  50.  this keeps the winc.rc file from influencing the results.

	+ winc.1: a few more comments about creating the starfield image.
	  the home camera wanted -S 20, the office camera wanted -S 45.
	  this is either a difference in cameras, or a difference in
	  temperature -- the office camera is on top of a warm monitor,
	  the one at home beside a cool wall.
    
Changes for winc-33:
    960828
	+ pnm.c image.c convert.c find.c private.h winc.c winc.1:  changes
	  to support starfield reference images.

	+ Makefile starfield.c: new file containing code for reading and using
	  the starfield reference image.

Changes for winc-32:
    960826
	+ wincd.c winc.c wincd.1 winc.1:  reset the mod time of the written
	  image to be the time at which the image was snapped.

	+ winc.c: added "-o filename" to replace stdout as image destination.

    960823
	+ winc.c: added -S, for saving starfield pbm file.  nothing uses
	  this yet.

	+ pnm.c pnm.h: added winc_pbm_output(), for mono bitmap output

Changes for winc-31: (mostly camera id support -- for naming cameras)
    960822
	+ winc.c: usage message is now split in two -- it was way
	  too long.

	+ winc.c: added -s speed option for setting baud rate.

	+ winc.1: rearranged options, added section on baud rate and
	  timeouts.

	+ grabber.sh: add disclaimer about how wincd/wincd_script do
	  everything (?) this does, perhaps better.

    960821
	+ convert.c winc.1:  added WinCamColor variable for specifying
	  default color balance.

	+ winc.1 wincd.1 xwinc.1: documentation of -i ID option.

    	+ winc.c: unlock camera before normal closing, to suppress
	  superfluous (and time-consuming) flush.

	+ config.c: add support $WinCamId as a prefix for other config
	  variables.  i.e. if WinCamId is set to "kitchen", then when
	  we are asked for to look up "WinCamDevice", we will check for
	  "kitchenWinCamDevice" first.

	+ mergeenv.c: added "safeputenv()", which mallocs before doing
	  a putenv() on its args.

	+ winc.c wincd.c xwinc.c: add support for "-i idstring", for
	  setting the WinCamId variable.  this lets you use multiple
	  cameras on the same system, all configured from the same
	  config file.

    	+ find.c: include string.h for strerror() decl.

    	+ convert.c: log the color balance choices made.

Changes for winc-30:  (mostly modem support, also wincd bugfix)
    960821
	+ find.c: if WinCamModemChat config variable is non-null, then
	  the line is tested for a modem.  if one is found, the contents
	  of that config variable are run via "sh -c", with the tty port
	  as stdin and stdout.  the command is intended to be a chat
	  program, which will negotiate with and dial the modem.

	+ protocol.c, private.h: added winc_check_for_modem() routine,
	  to distinguish modem ports from camera ports.

	+ comms.c: open the tty O_NDELAY, to ensure it opens.

	+ wincd.c:  change DEFAULT_MAXDIFF to just 1% (from 4%).  i
	  mis-translated the old numbers when i converted the code to use
	  percentages.  it _used_ to use just 10 out of 3072 (64x48)
	  pixels, which is .0032, or .32%, not 3.2%.

    960818
	+ wincd_script: change from /bin/sh to /bin/bash.

	+ winc.1 wincd.1: fix syntax error.

Changes for winc-29:	(mostly implementation of color-balance controls)

    960818
    	+ winc.c winc.1: no more -m option.  -e or -E forces manual
	  exposure, -b allows setting goal, which is passed as negative
	  exposure setting to the snap routines.

	+ image.c convert.c wincam.h view.c image.c:  exposure is passed as
	  int not unsigned, so goal can come in as negative exposure value.

	+ view.c: change default brightness to 50%.

	+ wincd_script:  make it a little smarter about its args.

	+ xwinc.c: brightness default now comes from config file.

	+ rationalize different default brightness settings, add -b flag
	  to winc.c for controlling brightness goal.

    	+ color.c: the problem that green-level is "off by 59%" is
	  now taken care of in here.  blue, red, and green fix values
	  are now all based on 128 externally.

	+ convert.c winc.c wincd.c wincam.c:  API change: added "image
	  adjustment" struct to arg list, added "winc_image_adjustments()"
	  routine to fill it in, based on user-provided color-balance string,
	  and widen and vignette choices.

	+ winc.c wincd.c winc.1 wincd.1:  added -l option, which lets
	  one set lighting (color-balance) specifications.

	+ winc.c: re-enabled the widen (-n) and vignette (-r) options.

    960817
    	+ convert.c: commentary about color balancing values.

	+ wincd.c wincd.1: change -l SECS to -p SECS (for "poll") to
	  free up -l for "lighting" control.

Changes for winc-28:	(mostly cosmetic renaming/relayering)
    960816
	+ comms.c config.c convert.c diags.c find.c locks.c private.h
	  protocol.c view.c vignette.c winc.1 wincam.h wincd.c xwinc.c:
	  mostly cosmetic.  renamed some internal routines to show their

src/winc.c

/*
 *
 * winc -- a command line program for accessing the WinCam.One digital camera.
 *
 * Copyright (C) 1996, Paul G. Fox
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the
 * Free Software Foundation; either version 2 of the License, or (at your
 * option) any later version.
 * 
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 * This software was created with the help of proprietary information
 * belonging to StarDot Technologies.
 *
 * $Header: E:/winc/RCS/winc.c 1.1 1997/03/01 03:44:14 Derek Exp Derek $
 */

#include <unistd.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <signal.h>
#include <time.h>
#include <utime.h>
#include <sys/types.h>
#include <sys/stat.h>
#include "wincam.h"
#include "pnm.h"
#include "trace.h"

/*
 * from mergeenv.c -- there's no convenient header file in which
 * this appears.
 */
int	safeputenv(char *name, char *value);

/*
 * forward declarations
 */
void emit_wcd_format(unsigned char *, int, unsigned char *, FILE *);
void inhale_wcd_format(unsigned char *, int *, FILE *);
void snapper_func(void *timep);
void audible_snapper_func(void *timep);
void reset_mod_time(char *filename, time_t mtime);
void quitcatcher(int sig);

/*
 * global camera handle, so it can be closed/flushed from a signal handler
 */
cam_t Cam = 0;		/* camera "handle" */

/*
 * for getopt()
 */
extern char *optarg;
extern int optind, opterr, optopt;

void
usage(char *progname, int full)
{
    fprintf(stderr, 
    "usage: %s -{flags}  where:\n%s%s", progname,
    "   -h gives the usage message\n"
    "   -H gives a more complete usage message\n"
    "   -c grabs a color image\n"
    "   -g grabs a greyscale image\n"
    "   -v grabs a viewfinder image\n"
    "     (if none of -v, -g, or -c is used, no image will be grabbed,\n"
    "      but camera will still be discovered and initialized)\n"
    "   -2 to double-scan an image\n"
    "   -z N sets image reduction factor, or viewfinder zoom. (1, 2, 4, or 8)\n"
    "   -b NN - brightness (1-100, 0 for default) [50]\n"
    "   -l color-adjustment\n"
    "   -a causes an audible alert (beep) when image is exposed\n"
    "   -s SPEED sets the preferred baud rate\n"
    "   -i IDSTRING assigns a prefix for all config info looked up\n"
    "   -o FILE sets the output file to use in place of stdout\n"
    "", !full ? "" :
    "   -S NN saves a starfield image, using NN as the \"bad\" threshold\n"
    "   -e NN sets exposure in milliseconds\n"
    "   -E NN sets exposure in microseconds\n"
    "   -f don't do faster grab\n"
    "   -n leaves \"narrow\" image, rather than widening\n"
    "   -r leaves a dark image, rather than brightening corners\n"
    "	-G VAL processes with gamma function . (100 normal, 110 default)\n"
    "   -d runs camera diagnostics\n"
    "   -t STR sets tracing; arg is a string of traceflags:\n"
    "	    v just shows program and camera version information\n"
    "	    V will turn on all possible messages\n"
    "	    e to get exposure information\n"
    "	    a for simple API tracing\n"
    "	    A for full API tracing\n"
    "	    c for camera commands\n"
    "	    s for serial protocol\n"
    "	    S for all bytes of serial protocol\n"
    "	    g for graphics output routine tracing\n"
    "	    r for winc.rc file tracing\n"
    "   -R writes .wcd format data\n"
    "   -D gets .wcd format data from stdin\n"
    "\n"
    );
}

int
main(int argc, char *argv[])
{
    char *progname;
    int c;


    int do_viewfinder = 0;	/* flags three kinds of grabs */
    int do_color = 0;
    int do_greyscale = 0;
    int do_starfield = 0;
    int do_raw_output = 0, do_raw_input = 0;
    int scantype = 1;		/* images can be single- or double- scanned */
    int fraction = 1;		/* full image? what fraction? */
    int widen = 1;		/* should we widen to fix aspect ratio? */
    int vignette = 1;		/* should we fix corner darkening? */
    int gamma = 0;		/* should we do darkness lightening? */
    int faster = 1;
    int audible_alert = 0;
    int starfield_cutoff = 1;

    int goal = 0;		/* brightness goal */
    int exp = 0;		/* default exposure is 1/30th of a second */
    int zoom = 1;		/* no viewfinder zoom by default */
    int diagnostics = 0;
    struct winc_image_adjust image_adjust;
    char *color_adj_string = 0;

    char *filename = 0;		/* where to put the image if not stdout */
    FILE *outf;			/* file handle for output */
    time_t snaptime;		/* for resetting image grab time */

    int rows, cols;
    int blackadj;		/* average black value */
    unsigned char blackvals[2*512 + 2]; /* space for black values, if needed  */
    unsigned char image[640*492];	/* the biggest imagemap we'll need */
    unsigned char convimage[3*640*492]; /* the biggest converted imagemap */


    /* 
     * what program is this?
     */
    progname = strrchr(argv[0], '/');
    if (progname)
	progname++;
    else
	progname = argv[0];

    _fsetmode(stdout,"b");

    /* 
     * set up the trace and error facilities
     * register the output callbacks, and initialize the tracing class
     */
    register_error_func((vprintffunc)vfprintf, (void *)stderr);
    register_trace_func((vprintffunc)vfprintf, (void *)stderr);
    set_trace_class("");

    /* 
     * process arguments
     */
    while ((c = getopt(argc, argv, 
		    "nr2hHvfgcdmaRDo:l:G:z:b:e:E:t:f:i:s:S:"
		    )) != EOF) {
	switch (c) {
	case 'H':
	case 'h':
	    usage(progname, (c == 'H'));
	    exit(0);
	case 'e':   /* exposure (or initial exposure guess), in milliseconds */
	case 'E':   /*	or microseconds */
	    exp = atoi(optarg);
	    if (c == 'e')
		exp *= 1000;	    /* convert to microseconds */
	    break;
	case 'n':
	    widen = 0;
	    break;
	case 'b':
	    goal = atoi(optarg);
	    break;
	case 'S':
	    do_starfield = 1;
	    scantype = 2;
	    goal = 50;
	    starfield_cutoff = atoi(optarg);
	    if (safeputenv("NoStarfield", "none_at_all") == -1) {
		fprintf(stderr, "%s: couldn't set NoStarfield\n",
			progname);
		exit(1);
	    }
	    break;
	case 'r':
	    vignette = 0;
	    gamma = 100;	/* force winc_gamma() to be transparent */
	    break;
	case 'G':
	    gamma = atoi(optarg);
	    break;
	case 'z':
	    zoom = atoi(optarg);
	    fraction = zoom;
	    break;
	case 'd':
	    diagnostics = 1;
	    break;
	case 't':
	    set_trace_class(optarg);
	    break;
	case 'a':
	    audible_alert  = 1;
	    break;
	case 'v':
	    do_viewfinder = 1;
	    break;
	case 'f':
	    faster = 0;
	    break;
	case 'g':
	    do_greyscale = 1;
	    break;
	case '2':
	    scantype = 2;
	    break;
	case 'c':
	    do_color = 1;
	    break;
	case 'l':
	    color_adj_string = optarg;
	    break;
	case 'i':
	    if (safeputenv("WinCamId", optarg) == -1) {
		fprintf(stderr, "%s: couldn't set WinCamId\n", progname);
		exit(1);
	    }
	    break;
	case 'o':
	    filename = optarg;
	    break;
	case 's':
	    if (safeputenv("WinCamBaudRate", optarg) == -1) {
		fprintf(stderr, "%s: couldn't set WinCamBaudRate\n", progname);
		exit(1);
	    }
	    break;
	case 'R':
	    /* option given once, output stardot .wcd format,
		if twice, output raw image array */
	    do_raw_output++;
	    faster = 0;
	    break;
	case 'D':
	    /* if option is given once, read stardot .wcd format,
		else raw image array */
	    do_raw_input++;
	    faster = 0;
	    break;
	default:
	    fprintf(stderr, "%s: unknown option '%c'\n", progname, c);
	    usage(progname, 0);
	    exit(1);
	}
    }

    if (optind != argc) {
	fprintf(stderr, "%s: extra arguments\n", progname);
	usage(progname, 0);
	exit(1);
    }

    if (do_viewfinder +
	do_greyscale +
	do_color +
	do_starfield +
	(do_raw_output != 0) > 1) {
	fprintf(stderr, 
	    "%s: only one of -v, -g, -c, -S, or -R should be given\n",
						progname);
	usage(progname, 0);
	exit(1);
    }

    if (filename) {
	outf = fopen(filename, "wb");
	if (!outf) {
	    errormsg(__FILE__ ": can't open file \"%s\"\n", filename);
	    exit(1);
	}
    } else {
	outf = stdout;
    }

    if (audible_alert != 0)
	winc_register_alerter(audible_snapper_func, (void *)&snaptime);
    else
	winc_register_alerter(snapper_func, (void *)&snaptime);

    winc_image_adjustments(&image_adjust, color_adj_string, widen, vignette);

    if (exp)
	goal = 0;	/* goal is meaningless if manual exposure */

    /* don't actually touch the camera if we're just debugging conversions */
    if (!do_raw_input) {
	/*
	 * finally, find a camera.  the argument (1) currently doesn't
	 * do anything -- eventually we should support more than one camera,
	 * on separate serial ports.
	 */
	Cam = winc_find(1);
	if (!Cam) {
	    fprintf(stderr, "%s: couldn't find camera\n", progname);
	    exit(1);
	}

	(void) signal(SIGHUP, quitcatcher);
	(void) signal(SIGINT, quitcatcher);
	(void) signal(SIGTERM, quitcatcher);


	if (diagnostics)
	    winc_diagnose(Cam);

	if (do_viewfinder) {

	    if (winc_viewfinder_snap(Cam, zoom, -1, -1, image,
			(exp == 0) ? 15 : 1, &exp, 0, goal, 0) != OK) {
		errormsg(__FILE__ ": failed to snap initial viewfinder image\n");
		exit(0);
	    }
	    /*
	     * viewfinder images are all 48 high and 64 across, and
	     * no massaging is necessary, even aspect ratio.
	     */
	    rows = 48;
	    cols = 64;
	    (void)winc_pgm_output(image, rows, cols, 15, outf);
	    winc_unlock(Cam);
	    winc_close(Cam);
	    exit(0);
	}
    }

    /* goal is passed as negative exposure, unless we're exposing manually */
    if (!exp && goal)
	exp = -goal;

    if (faster && (do_greyscale || do_color)) {
	winc_snap_and_process(Cam, exp, image, convimage, &rows, &cols,
				do_color, scantype, fraction, &image_adjust);
	winc_gamma(gamma, convimage, rows, cols, do_color);
	if (do_color)
	    (void)winc_pnm_output(convimage, rows, cols, 255, outf);
	else
	    (void)winc_pgm_output(convimage, rows, cols, 255, outf);
	if (filename && snaptime != 0) {
	    fclose(outf);
	    reset_mod_time(filename, snaptime);
	}
	winc_unlock(Cam);
	winc_close(Cam);
	return 0;
    }

    if (do_greyscale || do_color || do_starfield || do_raw_output) {
	if (!do_raw_input) {
	    if (winc_image_snap(Cam, exp, scantype, 
		    fraction, !do_greyscale, image, &rows, &cols) != OK) {
		errormsg(__FILE__ ": failed to snap image\n");
		winc_close(Cam);
		exit(0);
	    }

	    if (winc_black_offset(Cam, &blackadj, do_raw_output ? blackvals : 0 )
								    != OK) {
		errormsg(__FILE__ ": failed to get black adjustment\n");
		winc_close(Cam);
		exit(0);
	    }
	} else {
	    int ch;
	    unsigned char *ip = image;

	    if (do_raw_input > 1) {
		/* load up the data */
		while((ch = getc(stdin)) != EOF)
		    *ip++ = ch;

		if (scantype == 2) {
		    rows = 492; cols = 512;
		} else {
		    switch(fraction) {
		    case 1: rows = 246; cols = 512; break;
		    case 2: rows = 246; cols = 256; break;
		    case 4: rows = 123; cols = 128; break;
		    default:
		    case 8: rows = 60;	cols = 64;  break;
		    }
		}

		blackadj = 0; /* can't get this back easily */
	    } else {
		inhale_wcd_format(image, &blackadj, stdin);
		rows = 492;
		cols = 512;
		scantype = 2;
	    }
	}

#if USE_MY_GREY
	if (do_greyscale) {
	    winc_grey_convert(image, convimage, &rows, &cols, blackadj,
				scantype, fraction, &image_adjust);
	    if (vignette)
		winc_vignette_fix(convimage, rows, cols, 0);
	    if (widen)
		cols = winc_aspect_fix(convimage, rows, cols, 0);
	    (void)winc_pgm_output(convimage, rows, cols, 255, outf);
	} else if (do_color) {
	    winc_color_convert(image, convimage, &rows, &cols, blackadj,
				do_color, scantype, fraction, &image_adjust);
	    if (vignette)
		winc_vignette_fix(convimage, rows, cols, do_color);
	    if (widen)
		cols = winc_aspect_fix(convimage, rows, cols, do_color);
	    if (do_color)
		(void)winc_pnm_output(convimage, rows, cols, 255, outf);
	    else
		(void)winc_pgm_output(convimage, rows, cols, 255, outf);
	} 
#else
	if (do_color || do_greyscale) {
	    winc_color_convert(image, convimage, &rows, &cols, blackadj,
				do_color, scantype, fraction, &image_adjust);
	    winc_gamma(gamma, convimage, rows, cols, do_color);
	    if (do_color)
		(void)winc_pnm_output(convimage, rows, cols, 255, outf);
	    else
		(void)winc_pgm_output(convimage, rows, cols, 255, outf);
	} 
#endif
	else if (do_raw_output) {
	    if (do_raw_output > 1) {
		/* simple write of the image array, suitable for sucking
		    in with -D */
		(void)fwrite(image, 1, rows*cols, outf);
	    } else {
		emit_wcd_format(image, scantype, blackvals, outf);
	    }
	} else if (do_starfield) {
	    (void)winc_pbm_output(image, convimage, rows, cols,
	    		starfield_cutoff, outf);
	}
	if (filename && snaptime != 0) {
	    fclose(outf);
	    reset_mod_time(filename, snaptime);
	}
    }

    if (Cam) {
        winc_unlock(Cam);
	winc_close(Cam);
    }

    return 0;
}


/* 
 * two routines to read and write  StarDot's .wcd format:
 *  246 even rows (they call them odd; historical)
 *  10	rows of zero
 *  246 odd rows (again, they call these even)
 *  8	rows of zero
 *  2	rows of black values
 * all rows are 512 bytes across
 */

void
emit_wcd_format(
    unsigned char *image,
    int scantype,
    unsigned char *blackvals,
    FILE *outf)
{
    /* this won't contain interesting data if the scan
     *	wasn't a full height scan
     */
    unsigned char empty[512];
    unsigned char *p;
    int i;
    memset(empty, 0, sizeof(empty));

    /* write the even rows */
    p = image;
    for (i = 0; i < 246; i++) {
	(void)fwrite(p, 1, 512, outf);
	p += 512;
	if (scantype == 2)
	    p += 512;
    }

    /* ten empty rows */
    for (i = 0; i < 10; i++)
	(void)fwrite(empty, 1, 512, outf);

    /* write the odd rows, or write the even rows again */
    p = (scantype == 2) ? &image[512] : image;
    for (i = 0; i < 246; i++) {
	(void)fwrite(p, 1, 512, outf);
	p += 512;
	if (scantype == 2)
	    p += 512;
    }

    /* eight empty rows */
    for (i = 0; i < 8; i++)
	(void)fwrite(empty, 1, 512, outf);

    /* two black rows */
    (void)fwrite(blackvals, 1, 2*512, outf);
}

void
inhale_wcd_format(
    unsigned char *image,
    int *blackadjp,
    FILE *inf)
{
    int i, sum;
    unsigned char *p;
    unsigned char scratch[512];

    /* read the even rows */
    p = image;
    for (i = 0; i < 246; i++) {
	if (fread(p, 1, 512, inf) != 512)
	    goto errout;
	p += 2*512;
    }

    /* ten empty rows */
    for (i = 0; i < 10; i++) {
	if (fread(scratch, 1, 512, inf) != 512)
	    goto errout;
    }

    /* read the odd rows */
    p = &image[512];
    for (i = 0; i < 246; i++) {
	if (fread(p, 1, 512, inf) != 512)
	    goto errout;
	p += 2*512;
    }

    /* eight empty rows */
    for (i = 0; i < 8; i++) {
	if (fread(scratch, 1, 512, inf) != 512)
	    goto errout;
    }


    /* two black rows to average */
    sum = 0;
    if (fread(scratch, 1, 512, inf) != 512)
	    goto errout;
    for (i = 0; i < 246; i++)
	sum += scratch[i];

    if (fread(scratch, 1, 512, inf) != 512)
	    goto errout;
    for (i = 0; i < 246; i++)
	sum += scratch[i];

    sum /= (2 * 246);
    *blackadjp = sum;

    return;

 errout:
    fprintf(stderr, "Error reading .wcd format input\n");
    exit(1);
}


/*
 * the snapper funcs are "called back" by the winc library when the
 * picture is actually snapped.  this lets us a) make a noise, and b)
 * record the time
 */
void
snapper_func(void *timep)
{
    *(time_t *)timep = time(0);
}

void
audible_snapper_func(void *timep)
{
    putc('\a', stderr);
    *(time_t *)timep = time(0);
}

/*
 * attempt to reset the modification time of the given file to mtime.
 * this is used to set the time on the output file to the time at which
 * the image was snapped, rather than when it was written, since processing
 * can take some time.
 */
void 
reset_mod_time(char *filename, time_t mtime)
{
    struct stat stb;
    struct utimbuf utb;
    if (stat(filename, &stb) == -1)
    	return;

    utb.actime = stb.st_atime;
    utb.modtime = mtime;
    (void)utime(filename, &utb);
}

void
quitcatcher(int sig)
{
    if (Cam)
    	winc_close(Cam);
    exit(1);
}

